File: tests/refs/n_queens_count.py
"""Reference: count of distinct n-queens placements on an n x n board."""

from __future__ import annotations


def nQueensCount(n: int) -> int:
    count = 0
    queens = [-1] * n

    def is_safe(row: int, col: int) -> bool:
        for r in range(row):
            c = queens[r]
            if c == col or abs(c - col) == abs(r - row):
                return False
        return True

    def go(row: int) -> None:
        nonlocal count
        if row == n:
            count += 1
            return
        for col in range(n):
            if is_safe(row, col):
                queens[row] = col
                go(row + 1)
                queens[row] = -1

    if n > 0:
        go(0)
    return count

File: tests/refs/NQueensCount.java
/*
 * Reference Java implementation for tests/cases/n_queens_count.json.
 * Backtracking count of distinct n-queens placements.
 */
public class NQueensCount {
    private static int[] queens;
    private static int count;
    private static int N;

    private static boolean isSafe(int row, int col) {
        for (int r = 0; r < row; ++r) {
            int c = queens[r];
            if (c == col || Math.abs(c - col) == Math.abs(r - row)) return false;
        }
        return true;
    }

    private static void go(int row) {
        if (row == N) { count++; return; }
        for (int col = 0; col < N; ++col) {
            if (isSafe(row, col)) {
                queens[row] = col;
                go(row + 1);
                queens[row] = -1;
            }
        }
    }

    public static long nQueensCount(long n) {
        N = (int) n;
        count = 0;
        queens = new int[N];
        for (int i = 0; i < N; ++i) queens[i] = -1;
        if (N > 0) go(0);
        return count;
    }
}
